cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -66,6 +66,25 @@ struct BinaryLength {
 
 #ifdef ARROW_WITH_UTF8PROC
 
+struct Utf8Length {
+  template <typename OutValue, typename Arg0Value = util::string_view>
+  static OutValue Call(KernelContext*, Arg0Value val) {
+    auto str = reinterpret_cast<const utf8proc_uint8_t*>(val.data());
+    auto strlen = static_cast<utf8proc_ssize_t>(val.size());
+    utf8proc_int32_t codepoint;
+
+    OutValue length = 0;
+    while (strlen > 0) {
+      auto char_width = utf8proc_iterate(str, strlen, &codepoint);
```

Comment:
You may want to add a benchmark for this kernel in `scalar_string_benchmark.cc`. Then you can compare the utf8proc-based implementation against a more optimized implementation that will simply count of number of bytes smaller than 128.

Reply:
I do not think it is worthwhile, given that the optimized implementation consists of a pair of logical operations compared to [utf8proc_iterate](https://github.com/JuliaStrings/utf8proc/blob/master/utf8proc.c#L124-L171) which has several branches and a few other operations for validating continuation bytes. The rest of *utf8_length* implementation is basically the same for both variants, so the optimized version should be clearly faster.